File: SumORSubSet.go
package main

import (
	"fmt"
)

const BIT_SIZE int = 32

func getBit(n uint32, k uint32) uint32 {
	var bitSet uint32
	if n&(1<<k) != 0 {
		bitSet = 1
	}
	return bitSet
}

func bitWiseOR(arr []uint32) uint32 {
	var result uint32
	n := len(arr)
	bitFreq := make([]uint32, BIT_SIZE)
	for i := 0; i < n; i++ {
		for k := BIT_SIZE - 1; k >= 0; k-- {
			bitFreq[k] += getBit(arr[i], uint32(k))
		}
	}
	for j := 0; j < BIT_SIZE; j++ {
		for bitFreq[j] > 0 {
			result += ((1 << uint32(j)) * (1 << (uint32(n) - bitFreq[j])))
			bitFreq[j]--
		}
	}
	return result
}
func main() {
	arr := []uint32{1, 2, 3, 4}
	fmt.Println("print array:", arr)
	fmt.Println("bitWiseOr:", bitWiseOR(arr))

}

File: SumORSubSet.rs
const BIT_SIZE: usize = 32;
fn get_bit(n:i32, k:usize) -> i32{
    if n & (1<<k) !=0{
        return 1
    }
    return 0
}
fn bit_wise_or(arr: &[i32]) -> i32{
    let mut bit_freq = [0; BIT_SIZE]; //Array of BIT_SIZE element
    let n = arr.len();
    let mut result = 0; 
    for i in 0..n {
        for k in (0usize .. BIT_SIZE).rev(){
            bit_freq[k] += get_bit(arr[i], k);
        }
    }
    for j in (0usize .. BIT_SIZE).rev(){
        while bit_freq[j] > 0 {
            result += (1 << j) * (1 << (n as i32- bit_freq[j]));
            bit_freq[j]-=1;
        }
     }
     result
}

fn main() {
    let arr: [i32; 3] = [1, 5, 6];
    println!("array: {:?}", arr);
    println!("{}", bit_wise_or(&arr));
}
